[JSC] ArrayBuffer/SharedArrayBuffer constructor should check length before creating an instance

sosukesuzuki opened the pull request:
#### 2b5ed009bee397beb4099f6ea0e40072d2aaec0f
<pre>
[JSC] ArrayBuffer/SharedArrayBuffer constructor should check length before creating an instance
<a href="https://bugs.webkit.org/show_bug.cgi?id=272809">https://bugs.webkit.org/show_bug.cgi?id=272809</a>

Reviewed by Darin Adler.

According to the specs[1][2], the constructors for ArrayBuffer and SharedArrayBuffer should check
if `byteLength &gt; maxByteLength` and throw a `RangeError` before creating an instance.

This patch changes to perform these checks before creating an instance.

`toTypedArrayIndex`, there is an observable side effect of throwing a `RangeError` when `length`
exceeds `MAX_ARRAY_BUFFER_SIZE`. So, this patch use `toNumber` instead when checking
`byteLength &gt; maxByteLength`.

[1]: <a href="https://tc39.es/ecma262/#sec-allocatearraybuffer">https://tc39.es/ecma262/#sec-allocatearraybuffer</a>
[2]: <a href="https://tc39.es/ecma262/#sec-allocatesharedarraybuffer">https://tc39.es/ecma262/#sec-allocatesharedarraybuffer</a>

* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/JSArrayBufferConstructor.cpp:
(JSC::JSGenericArrayBufferConstructor&lt;sharingMode&gt;::constructImpl):

Canonical link: <a href="https://commits.webkit.org/279181@main">https://commits.webkit.org/279181@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5c7174f25987ce1591fa518b90864ab0a4438ee0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52587 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31921 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5013 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55861 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3310 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/38491 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3011 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42724 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2119 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54685 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29601 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45384 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23818 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26747 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2665 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1469 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/45938 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48635 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2818 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57457 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/52097 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27722 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/2853 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50119 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28948 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45500 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49388 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11514 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29861 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/64403 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28698 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12207 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

shvaikalesh commented on Source/JavaScriptCore/runtime/JSArrayBufferConstructor.cpp:
```diff
@@ -104,8 +102,21 @@ EncodedJSValue JSGenericArrayBufferConstructor<sharingMode>::constructImpl(JSGlo
+
+    size_t length = 0;
+    if (hasArguments) {
+        length = callFrame->uncheckedArgument(0).toTypedArrayIndex(globalObject, "length"_s, lengthDouble);
```
Is there any chance we can avoid adding third parameter to `toTypedArrayIndex`? Maybe we can wrap `lengthDouble` in JSValue and call the method on it? It is very confusing to have an optional parameter that changes the method's operand.

sosukesuzuki replied:
Thank you for the review! I too would prefer to avoid adding a new argument if possible.

The problem is that `toNumber` has observable side effects, namely it calls `valueOf`. Also, `toTypedArrayIndex` internally calls `toNumber`.

By calling `toNumber` on `length` for the `length < maxByteLength` check, and then calling `toTypedArrayIndex`, we end up calling `toNumber` twice, resulting in `valueOf` being called twice. This behavior causes tests like the one at https://github.com/WebKit/WebKit/blob/284bb1401647bf26edf133c6543caa7f280e94d2/JSTests/stress/v8-resizablearraybuffer-growablesharedarraybuffer.js#L168-L177 to fail.

To avoid this problem, I've added a new argument to reuse the `double` value already obtained by `toNumber`...

sosukesuzuki replied:
Ahh sorry, I misunderstood your comment. I have addressed it.

sosukesuzuki replied:
@shvaikalesh friendly ping

sosukesuzuki committed:
[JSC] ArrayBuffer/SharedArrayBuffer constructor should check length before creating an instance
https://bugs.webkit.org/show_bug.cgi?id=272809

Reviewed by Darin Adler.

According to the specs[1][2], the constructors for ArrayBuffer and SharedArrayBuffer should check
if `byteLength > maxByteLength` and throw a `RangeError` before creating an instance.

This patch changes to perform these checks before creating an instance.

`toTypedArrayIndex`, there is an observable side effect of throwing a `RangeError` when `length`
exceeds `MAX_ARRAY_BUFFER_SIZE`. So, this patch use `toNumber` instead when checking
`byteLength > maxByteLength`.

[1]: https://tc39.es/ecma262/#sec-allocatearraybuffer
[2]: https://tc39.es/ecma262/#sec-allocatesharedarraybuffer

* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/JSArrayBufferConstructor.cpp:
(JSC::JSGenericArrayBufferConstructor<sharingMode>::constructImpl):

Canonical link: https://commits.webkit.org/279181@main
